worlds/sc2/MissionOrders.py

```diff
@@ -0,0 +1,389 @@
+    while mission_number < num_missions:
+        column = mission_number % width
+        row = math.floor(mission_number / width)
+        if mission_number == middle_column:
+            difficulty = MissionPools.STARTER
+        elif mission_number == final_mission_number:
+            difficulty = MissionPools.FINAL
+        else:
+            difficulty = smooth_difficulty[(min(row, max_difficulty))]
+        mission_order.append(FillMission(
+            difficulty,
+            connections,
+            f'_{column}',
+            or_requirements=True
+        ))
+        mission_number += 1
+        # Next row, requires previous row
+        if mission_number % width == 0:
+            connections = [MissionConnection(mission_number - 1 - i) for i in range(width)]
+            connections.reverse()
+    return {SC2Campaign.GLOBAL: mission_order}
+
+
+def make_hopscotch(two_start_positions: bool, num_missions: int):
+    mission_order: List[FillMission] = []
+    menu_connection = [MissionConnection(-1)]
+    max_width = 7
+    difficulty_progress = 0
+    difficulty_rate = max(0.5, min(1, 3 * max_difficulty / num_missions))
+    x = 0
+    y = 0
+    difficulty = MissionPools.EASY
+    first_diagonal = True
+    creation_cycle = 0  # Root -> Bottom Branch -> Right Branch
+    # Creating the starter missions
+    if two_start_positions:
+        mission_order += [
+            FillMission(MissionPools.STARTER, menu_connection, '_1', ui_vertical_padding=1),
+            FillMission(MissionPools.EASY, menu_connection, '_2')
+        ]
+        mission_number = 2
+    else:
+        mission_order.append(FillMission(MissionPools.STARTER, menu_connection, '_1', completion_critical=True))
+        creation_cycle = 1
+        mission_number = 1
+    while mission_number < num_missions:
+        if mission_number == num_missions - 1:
+            difficulty = MissionPools.FINAL
+        elif creation_cycle == 0 and difficulty_progress < max_difficulty:
+            difficulty_progress += difficulty_rate
+            difficulty = smooth_difficulty[math.floor(difficulty_progress)]
+        if creation_cycle == 0:
+            # Root
+            if y != -1:
+                x += 1
+            y += 1
+            mission_order.append(FillMission(
+                difficulty,
+                [MissionConnection(mission_number - 1), MissionConnection(mission_number - 2)],
+                f'_{x + 1}',
+                or_requirements=True,
+                completion_critical=True
+            ))
+        elif creation_cycle == 1:
+            # Bottom branch
+            mission_order.append(FillMission(
+                difficulty,
+                [MissionConnection(mission_number - 1)],
+                f'_{x + 1}'
+            ))
+        else:
+            # Right branch
+            column = f'_{x + 2}'
+            if x >= max_width - 1:
+                # Wrapping around
+                x = 0
+                y = -1
```

Comment:
It's possible to determine the x and y coordinates entirely from the `creation_step` and iteration number (where an iteration is a group of 3 missions being added). If this already works I'm not sure if it's worth refactoring it, though.

This also saves you from having to track `y`, as it's only used in this branch.

```python
STEP_ROOT = 0
STEP_BOTTOM = 1
STEP_RIGHT = 2

iteration_number = (mission_number + two_start_positions) // 3
x = (iteration_number + (creation_step == STEP_RIGHT)) % max_width
y = iteration_number + (creation_step == STEP_BOTTOM) - ((mission_number + 1 + two_start_positions) // (max_width * 3))
```

Reply:
I suppose when it comes to ease-of-reading, it's more about preference or style. As I said in the comment, it's probably not worth refactoring, so I'm cool to drop it. My perspective is that it's much easier to read and check correctness on a loop that has as few accumulators as possible, as accumulators must work on all branches/break/continue/early return, invariants between accumulators must be verified to match at every point, etc. Having 0 or one accumulators, and all values calculated from that is much easier to verify and maintain. Given you've already verified this works it's less of an issue.

Expectation-wise, as soon as I see something with cycles / wrapping around lines, I'm expecting everything to be in terms of modulo and integer-division calculations. Accumulators hide those things and spread out the logic, again making it harder to verify and preventing viewers from having simple and expected visual anchors. You could also think of all of these calculations as sampling sawtooth waves, which would keep periods, starting offsets, and sampling values all located together. It also makes it easier to optimize from there -- you can start at phase offset 1 (or 2 for two_start_positions), and all those pesky phase-offset calculations go away and everything is just an integer divide or module to get `x` or `y`.

Computationally-wise it's not much more expensive (at the least, it's all O(n) in the number of missions, so same complexity, while allowing for more conditional evaluation. As soon as you're comparing algorithms in the same big-O, actual performance gets really hard to judge as branches have variable cost, memory layout plays a humongous role, and all of this is invisible in Python). On modern processors I'm pretty sure modulo and integer division have the same cost cycle- and throughput- wise as integer add.

Reply:
I don't think myself this code needs a low-level optimization. We aren't doing in assembly. The code must be readable so it's intent is clear what is supposed to do. If you leave complex formula without explanation, it gets hard to get what that formula actually means later on.

Reply:
I don't think this code needs to be redone, it works after all. My initial comment was just to point out there was a way I'd consider better, so if Magne liked the concept / argument he could keep it in mind later. I maintain that having multiple accumulators in a loop is less readable, though, and the performance argument doesn't really hold a lot of weight in Python -- the big O is the same, mod instructions are not that much slower than add (we're talking like 9 clock cycles difference), and with the Python JIT overhead we're probably looking at fractions of a percent difference for one calculation. If you cared that much about that level of performance, I'd expect you to be using numpy, which would actually play nicer without accumulators as you could vectorize everything.